pkg/sql/importer/import_pgdump.go

```diff
@@ -0,0 +1,581 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/ioctx"
+	"github.com/cockroachdb/errors"
+)
+
+const (
+	// TempImportPGDumpDBName is the default name of the temp database for newly
+	// created schema from DDL stmts in the dump file.
+	TempImportPGDumpDBName = "crdb_temp_pgdump_import"
+)
+
+// createTempImportDatabase creates a temporary database where we will create
+// all the PGDUMP objects during the import.
+func createTempImportDatabase(ctx context.Context, p sql.JobExecContext) error {
+	id, err := p.ExecCfg().DescIDGenerator.GenerateUniqueDescID(ctx)
+	if err != nil {
+		return err
+	}
+	tempDBDesc := dbdesc.NewInitial(id, TempImportPGDumpDBName, username.NodeUserName())
+	return p.ExecCfg().InternalExecutorFactory.DescsTxn(ctx, p.ExecCfg().DB, func(
+		ctx context.Context, txn *kv.Txn, col *descs.Collection,
+	) error {
+		b := txn.NewBatch()
+		if err := col.WriteDescToBatch(
+			ctx, true /* kvTrace */, tempDBDesc, b,
+		); err != nil {
+			return err
+		}
+		b.CPut(catalogkeys.MakeDatabaseNameKey(p.ExecCfg().Codec, TempImportPGDumpDBName), tempDBDesc.GetID(), nil)
+		return txn.Run(ctx, b)
+	})
+}
+
+// PostgresDDLHandler stores DDL stmts and the temp db to store the new schemas.
+type PostgresDDLHandler struct {
+	dumpDatabaseName string
+	// TODO(adityamaru): Maybe memory monitor?
+	BufferedDDLStmts []string
+}
+
+func formatStatement(n tree.Statement) string {
+	f := tree.NewFmtCtx(
+		// TODO(adityamaru): should this be serializable?
+		tree.FmtParsable,
+	)
+	f.FormatNode(n)
+	return f.CloseAndGetString()
+}
+
+func rewriteStatementTableName(n tree.Statement, dumpDatabaseName string) (string, error) {
+	var err error
+	f := tree.NewFmtCtx(
+		// TODO(adityamaru): should this be serializable?
+		tree.FmtParsable,
+		tree.FmtReformatTableNames(func(fmtctx *tree.FmtCtx, tn *tree.TableName) {
+			// If the node has an explicit catalog name then we must replace it with
+			// the temporary database we are importing into.
+			// We only allow "db.schema.table" or "schema.table" form.
+			if tn.CatalogName != "" {
+				if checkCatNameErr := checkCatalogName(tn.CatalogName, tree.Name(dumpDatabaseName)); checkCatNameErr != nil {
+					err = checkCatNameErr
+					return
+				}
+				tn.CatalogName = TempImportPGDumpDBName
+			}
+			fmtctx.WithReformatTableNames(nil, func() {
+				fmtctx.FormatNode(tn)
+			})
+		}),
+	)
+	if err != nil {
+		return "", err
+	}
+	f.FormatNode(n)
+	return f.CloseAndGetString(), nil
+}
+
+func bufferDDLPostgresStatement(
+	ctx context.Context,
+	evalCtx *eval.Context,
+	postgresStmt tree.Statement,
+	p sql.JobExecContext,
+	parentID descpb.ID,
+	handler *PostgresDDLHandler,
+) error {
```

Comment:
Should this function or its parent check for context cancellation? I ask because I imagine that pgdumps might be quite big and take significant time to process.

In a similar vein, should we log the statements? Perhaps only when `log.ExpensiveLogEnabled`?